### PR TITLE
Add an example showing Cluster Sharding

### PR DESCRIPTION
**Pull request by raboof**
Showing cluster sharding in its most basic form, without persistence.

I created this example mainly to more easily test clusters that consist
of nodes with and without the 'cluster teams' feature, but might as well
have it here. If this example seems reasonable I'll also add a Java version.

### CODE REVIEW

**Comment by johanandren** [comment]
I have mixed feelings, on one hand good with a simple example, on the other hand this could (and will) be interpreted as us saying that this is a good way to do counter-per-id which it obviously isn't really. I'd wish for a more reasonable entity, but can't really come up with something from the top of my head.

**Comment by raboof** [comment]
Hmm, I agree we should be explicit about which parts are really what we recommend people doing, and which are just placeholders to make the example work.

For example I think we've seen some people use the `Main` from the samples as starting point for their production applications, even starting multiple `ActorSystems` from there..

Perhaps a couple of comments making that explicit would be sufficient?

File: akka-sample-sharding-scala/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+This tutorial contains a sample illustrating [Akka Cluster Sharding](http://doc.akka.io/docs/akka/current/scala/cluster-sharding.html#an-example).
+
+## Example overview
+
+First of all, make sure the correct settings in [application.conf](src/main/resources/application.conf) are set as described in the akka-sample-cluster tutorial.
+
+Open [ShardingApp.scala](src/main/scala/sample/sharding/ShardingApp.scala).
+
+This small program starts an ActorSystem with Cluster Sharding enabled.It joins the cluster and starts a `Counters` actor. This actor starts the infrastructure to shard `Counter` instances and starts sending messages to arbitrary counters.
```
Comment (patriknw) [inline akka-sample-sharding-scala/README.md]: missing space "enabled. It"

**Comment by patriknw** [comment]
Wouldn't a cool domain be to use the `Device` actors from the Getting Started tutorial. It even makes somewhat sense that they are not persistent because new temperature observations will arrive.

File: akka-sample-sharding-scala/src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      "akka.tcp://ShardingSystem@127.0.0.1:2552"]
+
+    # auto downing is NOT safe for production deployments.
+    # you may want to use it during development, read more about it in the docs.
+    auto-down-unreachable-after = 10s
+  }
+}
+
+# Disable legacy metrics in akka-cluster.
+akka.cluster.metrics.enabled=off
```
Comment (patriknw) [inline akka-sample-sharding-scala/src/main/resources/application.conf]: I think we removed that in 2.5, so shouldn't be needed any more

File: akka-sample-sharding-scala/src/main/resources/application.conf [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+# Enable metrics extension in akka-cluster-metrics.
+akka.extensions=["akka.cluster.metrics.ClusterMetricsExtension"]
```
Comment (patriknw) [inline akka-sample-sharding-scala/src/main/resources/application.conf]: don't enable metrics in this example

**Comment by raboof** [comment]
(logged test failure as #31)

**Comment by raboof** [comment]
`Device` seems like a neat domain. I'd like to keep some state to drive home the point that that state gets lost when the actor is migrated.

Perhaps let's send devices `RecordTemperature(deviceId, temperature)` messages and log the 'average temperature so far'?

**Comment by patriknw** [comment]
state can be latest temperature recording, but average is also good

File: akka-sample-sharding-scala/src/main/scala/sample/sharding/Devices.scala [hunk 4]
```diff
@@ -7,48 +7,45 @@ import scala.util.Random
...
 
   def receive = {
-    case UpdateCounter => 
-      val counterId = random.nextInt(numberOfCounters);
-      val msg = 
-        if (random.nextBoolean()) Counter.Increment(counterId) 
-        else Counter.Decrement(counterId)
+    case UpdateDevice =>
+      val deviceId = random.nextInt(numberOfDevices)
+      val temperature = random.nextDouble() * 100
```
Comment (patriknw) [inline akka-sample-sharding-scala/src/main/scala/sample/sharding/Devices.scala]: perhaps more realistic: `15.0 + 10 * random.nextDouble()`

**Comment by raboof** [comment]
I'll add the Java variant in a separate PR

**Pull request merged**